graphql_compiler/schema/__init__.py

```diff
@@ -389,10 +400,11 @@ def _parse_datetime_value(value: Any) -> datetime:
 GraphQLDateTime = GraphQLScalarType(
     name="DateTime",
     description=(
-        "The `DateTime` scalar type represents timezone-naive second-accuracy timestamps."
+        "The `DateTime` scalar type represents timezone-naive timestamps with microsecond accuracy."
         "Values are serialized following the ISO-8601 datetime format specification, "
-        'for example "2017-03-21T12:34:56". All of these fields must be included, '
-        "including the seconds, and the format followed exactly, or the behavior is undefined."
+        'for example "2017-03-21T12:34:56.012345" or "2017-03-21T12:34:56". All fields down to '
+        "and including seconds must be included, while fractional seconds are optional."
+        "If this format is not followed, the behavior is undefined."
```

Comment:
Is there a compelling reason to leave this behavior undefined?  Strict adherence to ISO-8601 isn't particularly onerous, and I'd much rather we just reject deviations outright than interpret them in undefined ways.

Reply:
Honestly, largely because I don't trust date/datetime parsing libraries to not get "too creative" in how they interpret dates. If we specify a particular behavior, then we are on the hook for any parsing bugs or inconsistencies in those libraries that happen _outside of_ ISO 8601 string cases.

For example, the `parse()` function on the GraphQL `DateTime` type can be called with any input data, including `int` and `bytes` and `float` and almost-but-not-quite ISO 8601, and while we have some best-effort tests for many of these cases, I don't want to overpromise nor have to own the API and test quality of the underlying libraries for anything outside of the "known to work" ISO 8601 seconds-included, fractionals-optional format we declare acceptable and extensively tested.

Reply:
I understand your argument, but I'm not convinced that a note in the docs is better than implementing our own strict parser--an important part of the compiler's value proposition is that it eliminates entire classes of weird surprises, so saying "don't do this or you might get weird surprises!" feels jarringly out of place to me.

Reply:
I would suggest to avoid implementing a separate ISO-8601 parser solely for this purpose. 

Hardly anyone needs support for all ISO-8601 formats, and the full spec is pretty challenging to implement (so no one in the python eco-system got it right as far as I know).

I would suggest offering support for 2-3 common formats (e.g., YYYY-MM-DDTHH:MM:SS and a few other variants), using arrow to parse those formats and checking that the length of the string being parsed matches the length of the format.

Reply:
Didn't see the thread above -- if ciso8601 works than that's great

Reply:
I agree that we don't need all of ISO-8601.

I'd be fine committing to `ciso8601`'s behavior, which has been stable for more than two years, and dropping the admonition about undefined behavior.  If @obi1kenobi or anyone else is uncomfortable about dropping the admonition when parsing is delegated to third-party code (which I think is a reasonable concern), then I suggest we implement a parser for RFC 3339's `date-time` and `full-date` rules, possibly with the following variance to [the RFC 3339 grammar](https://tools.ietf.org/html/rfc3339#section-5.6):
```
partial-time = time-hour ":" time-minute [time-second]
time-second  = ":" 2DIGIT [time-secfrac]
```

This wouldn't take much to implement, and I'd prefer it strongly to undefined behavior.  (Though, again, I'm fine just making `ciso8601`'s behavior part of our major-version contract.)

Reply:
`ciso8601` is probably both faster and more likely to be correct than a parser we write 😄 

How about we explicitly state which of the many ISO 8601 formats we support:
- `YYYY-MM-DD` for dates
- `YYYY-MM-DDThh:mm:ss`, optionally suffixed with up to 6 decimal places of fractional seconds `.ssssss`, for datetimes

And then use `ciso8601` instead of `arrow` to parse them and call it a day? 😄

Reply:
I don't think we can prevent `ciso8601` from parsing the other formats it supports, and this is where the undefined behavior in the spec is useful. We can guarantee parsing works for the formats we want to support, and everything else is left up to the implementation or completely undefined. It communicates that it's not a breaking change if we decide in the future to switch from `ciso8601` to another library that parses other "extras" in addition to the formats we promise to support.

I really don't think there's value in implementing our own parser, and I really don't think we stand to gain anything by making our spec tighter than "here are the formats we definitely support, we don't lose precision within them, everything else is undefined or implementation-defined."